pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go

```diff
@@ -28,4 +36,76 @@ type kubeEndpointsController interface {
 	InstancesByPort(c *Controller, svc *model.Service, reqSvcPort int,
 		labelsList labels.Collection) ([]*model.ServiceInstance, error)
 	GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance
+	updateEDS(obj interface{}, event model.Event)
+}
+
+type kubeEndpoints struct {
+	c        *Controller
+	informer cache.SharedIndexInformer
+}
+
+// GetProxyServiceInstances returns service instances of the given proxy.
+func (e *kubeEndpoints) GetProxyServiceInstances(c *Controller, proxy *model.Proxy, proxyNamespace string) []*model.ServiceInstance {
+	var otherNamespaceEndpoints []*model.ServiceInstance
+	sameNamespaceEndpoints := make([]*model.ServiceInstance, 0)
+
+	for _, item := range e.informer.GetStore().List() {
+		var proxyServiceInstances []*model.ServiceInstance
+		namespace := ""
+		switch item := item.(type) {
+		case *v1.Endpoints:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpoint(c, *item, proxy)
+		case *discoveryv1alpha1.EndpointSlice:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpointSlice(c, item, proxy)
+		}
+		if namespace == proxyNamespace {
+			sameNamespaceEndpoints = append(sameNamespaceEndpoints, proxyServiceInstances...)
+		} else {
+			if otherNamespaceEndpoints == nil {
+				otherNamespaceEndpoints = make([]*model.ServiceInstance, 0)
+			}
+			otherNamespaceEndpoints = append(otherNamespaceEndpoints, proxyServiceInstances...)
+		}
+	}
+
+	// Put the sameNamespaceEndpoints in front of otherNamespaceEndpoints so that Pilot will
+	// first use endpoints from sameNamespaceEndpoints. This makes sure if there are two endpoints
+	// referring to the same IP/port, the one in sameNamespaceEndpoints will be used. (The other one
+	// in otherNamespaceEndpoints will thus be rejected by Pilot).
+	return append(sameNamespaceEndpoints, otherNamespaceEndpoints...)
+}
+
+func (e *kubeEndpoints) HasSynced() bool {
+	return e.informer.HasSynced()
+}
+
+func (e *kubeEndpoints) Run(stopCh <-chan struct{}) {
+	e.informer.Run(stopCh)
+}
+
+func (e *kubeEndpoints) handleEvent(epc kubeEndpointsController, name string, namespace string, event model.Event, obj interface{}) error {
+	log.Debugf("Handle event %s for endpoint %s in namespace %s", event, name, namespace)
+
+	// headless service cluster discovery type is ORIGINAL_DST, we do not need update EDS.
+	if features.EnableHeadlessService.Get() {
+		if obj, _, _ := e.c.services.GetIndexer().GetByKey(kube.KeyFunc(name, namespace)); obj != nil {
```

Comment:
As far as I can tell this is not right. `name` here is the name of the endpoint object, right? It needs to be the name of the Service

Reply:
good catch, for endpointslice it is not right

Reply:
Yeah. thank. Will fix it.